### ISSUE TITLE
Enzyme does not recognize custom derivatives

### ISSUE THREAD

**Issue by oursland** [wsmoses/Enzyme#327]
## Description

Enzyme does not honor or recognize the custom derivatives registered via the `__enzyme_register_gradient*` struct.

## Configuration:

* OS: macOS Big Sur 11.5.2
* Enzyme: 0.0.18 (installed with Homebrew)
* LLVM: 12.0.1 (installed with Homebrew)

## Example Source

```C
#include <stdio.h>

double __enzyme_autodiff(void*, double);

void square_(const double* src, double* dest) {
    printf("%s()\n", __FUNCTION__);
    *dest = *src * *src;
}

void* augment_square_(const double* src, const double *d_src, double* dest, double* d_dest) {
    printf("%s()\n", __FUNCTION__);
    // intentionally incorrect for debugging
    *dest = 7.0;
    *d_dest = 11.0;
    return NULL;
}

void gradient_square_(const double* src, double *d_src, const double* dest, const double* d_dest, void** tape) {
    printf("%s()\n", __FUNCTION__);
    // intentionally incorrect for debugging
    *d_src = 13.0;
}

void* __enzyme_register_gradient_square[] = {
    (void*)square_,
    (void*)augment_square_,
    (void*)gradient_square_,
};


double square(double x) {
    printf("%s()\n", __FUNCTION__);
    double y;
    square_(&x, &y);
    return y;
}

double dsquare(double x) {
    printf("%s()\n", __FUNCTION__);
    return __enzyme_autodiff((void*)square, x);
}


int main() {
    double i = 3.0;
    printf("square(%f)=%f, dsquare(%f)=%f\n", i, square(i), i, dsquare(i));
}
```

## Expected Behavior

```
square()
square_()
dsquare()
square()
augment_square_() // <-- NOT CALLED
square_()
gradient_square_() // <-- NOT CALLED
square(3.000000)=9.000000, dsquare(3.000000)=13.000000 // <-- 13.0 to confirm derivative is from gradient_square_()
```

## Actual Behavior

```
square()
square_()
dsquare()
square()
square_()
square(3.000000)=9.000000, dsquare(3.000000)=6.000000
```

## Steps to Reproduce

1. Compile the example code with the command:
    `/usr/local/Cellar/llvm/12.0.1/bin/clang++ -std=c++17 -g -O1 test.cpp -o test -Xclang -load -Xclang /usr/local/Cellar/enzyme/0.0.18/lib/ClangEnzyme-12.dylib`
2.  Run `./test`

**Comment by wsmoses** [comment]
This should be resolved in #329. Note that at higher optimization levels you may need to mark the function as noinline (see the test in that PR), lest otherwise the call to the function will disappear.

**Comment by oursland** [comment]
Thanks for the fast response!  I'll test these changes out this evening.

**Comment by oursland** [comment]
Fix confirmed.  When merged, PR #329 will resolve this issue.
